### PR TITLE
TGBOT-279 Add Jacoco, SonarCloud and checkstyle

### PR DESCRIPTION
**Pull request by schneidervs**
https://schneidervs.atlassian.net/browse/TGBOT-279

### CODE REVIEW

**Review by yurasava — DISMISSED** [review]

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    types: [opened, synchronize, reopened]
+jobs:
+  build:
+    name: Build and analyze
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0  # Shallow clones should be disabled for a better relevancy of analysis
+      - name: Set up JDK 17
```
Comment (max-java) [inline .github/workflows/build.yml:16]: if we stay on v17, that should correspond in pom.xml 
Maybe 19?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        uses: actions/cache@v1
+        with:
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-m2
+      - name: Build and analyze
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}  # Needed to get PR information, if any
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+        run: mvn -B verify org.sonarsource.scanner.maven:sonar-maven-plugin:sonar -Dsonar.projectKey=Tutritp01_API-servise-TgBot
```
Comment (max-java) [inline .github/workflows/build.yml:36]: ass new line

Comment (max-java) [inline .github/workflows/build.yml:36]: Is it possible to encrypt this key?

File: build-tool/pom.xml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        <dependency>
+            <groupId>com.tutrit</groupId>
+            <artifactId>rest-api-service</artifactId>
+            <version>0.1.0-SNAPSHOT</version>
+        </dependency>
+    </dependencies>
+
+    <build>
+    <plugins>
+    <plugin>
```
Comment (max-java) [inline build-tool/pom.xml:38]: add tab alignment

File: build-tool/pom.xml [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        <configuration>
+            <systemPropertyVariables>
+                <jacoco-agent.destfile>**/jacoco.exec</jacoco-agent.destfile>
+            </systemPropertyVariables>
+        </configuration>
+    </plugin>
+    </plugins>
+    </build>
+
+</project>
```
Comment (max-java) [inline build-tool/pom.xml:112]: add new line

File: build-tool/pom.xml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    <build>
+    <plugins>
+    <plugin>
+        <groupId>org.jacoco</groupId>
+        <artifactId>jacoco-maven-plugin</artifactId>
+        <version>0.8.8</version>
+        <executions>
+            <execution>
+                <goals>
```
Comment (max-java) [inline build-tool/pom.xml:44]: Should this work for every module ? Maybe move this to parent pom.

File: build-tool/src/main/resources/config/checkstyle/checkstyle.xml [hunk 6]
```diff
@@ -0,0 +1,197 @@
+<?xml version="1.0"?>
```
Comment (max-java) [inline build-tool/src/main/resources/config/checkstyle/checkstyle.xml:1]: add checking for empty line in the end of a file

File: persistence-core/pom.xml [hunk 7]
```diff
@@ -10,7 +10,6 @@
         <version>0.1.0-SNAPSHOT</version>
     </parent>
 
-    <groupId>com.tutrit</groupId>
```
Comment (max-java) [inline persistence-core/pom.xml:13]: this change isn't related to issue 279

File: pom.xml [hunk 8]
```diff
@@ -18,8 +18,65 @@
         <!--        Added you module to this block-->
         <module>rest-api-service</module>
         <module>persistence-core</module>
+        <module>build-tool</module>
```
Comment (max-java) [inline pom.xml:21]: I suggest use jacoco, sonar and checstyle for all module, there for adding build-tool in the scope of issue 279 redundant

File: pom.xml [hunk 9]
```diff
@@ -18,8 +18,65 @@
...
+        <sonar.dynamicAnalysis>reuseReports</sonar.dynamicAnalysis>
+        <sonar.jacoco.reportPath>${project.basedir}\target\jacoco.exec</sonar.jacoco.reportPath>
+        <sonar.language>java</sonar.language>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-checkstyle-plugin</artifactId>
+            <version>3.2.1</version>
```
Comment (max-java) [inline pom.xml:38]: all versions should be moved to properties

File: pom.xml [hunk 10]
```diff
@@ -30,4 +87,216 @@
...
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-site-plugin</artifactId>
+                    <version>3.7.1</version>
+                </plugin>
+
+                <plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-project-info-reports-plugin</artifactId>
+                    <version>3.0.0</version>
+                </plugin>
```
Comment (max-java) [inline pom.xml:255]: I don't think we need to fix versions for this plugins, don't we ?

File: rest-api-service/pom.xml [hunk 11]
```diff
@@ -9,10 +9,8 @@
         <version>0.1.0-SNAPSHOT</version>
     </parent>
 
-    <groupId>com.tutrit</groupId>
     <artifactId>rest-api-service</artifactId>
     <version>0.1.0-SNAPSHOT</version>
-    <packaging>pom</packaging>
```
Comment (max-java) [inline rest-api-service/pom.xml:15]: this changes are not in the scope of #279 issue

**Review by max-java — CHANGES_REQUESTED** [review]
ptal

**Review by AlexanderYarokhovich — DISMISSED** [review]

**Review by Artem-Metlushko — DISMISSED** [review]

**Review by yurasava — APPROVED** [review]

**Review by max-java — APPROVED** [review]
lgfm

**Review by AlexanderYarokhovich — DISMISSED** [review]

**Review by Sav4ykVladzislau — DISMISSED** [review]

**Review by yurasava — DISMISSED** [review]

**Review by Sav4ykVladzislau — APPROVED** [review]

**Review by yurasava — APPROVED** [review]

File: integration/pom.xml [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+        <dependency>
+            <groupId>com.tutrit</groupId>
+            <artifactId>rest-api-service</artifactId>
+            <version>0.1.0-SNAPSHOT</version>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
```
Comment (max-java) [inline integration/pom.xml:30]: I think that could be inherited from parent pom, isn't it?

**Review by max-java — APPROVED** [review]
let's merge it and see how it works

File: integration/pom.xml (hunk 12 above)
Comment (schneidervs) [reply integration/pom.xml:30]: Yes

**Review by Artem-Metlushko — DISMISSED** [review]

**Review by AlexanderYarokhovich — APPROVED** [review]

**Review by Sav4ykVladzislau — DISMISSED** [review]

**Review by Sav4ykVladzislau — APPROVED** [review]

**Review by max-java — APPROVED** [review]
lgfm

**Review by yurasava — APPROVED** [review]

**Review by AlexanderYarokhovich — APPROVED** [review]

**Review by Artem-Metlushko — APPROVED** [review]

**Pull request merged**